python/plugins/processing/algs/qgis/PointDistance.py

```diff
@@ -225,27 +229,30 @@ def regularMatrix(self, parameters, context, source, inField, target_source, tar
                 break
 
             inGeom = inFeat.geometry()
-            inID = str(inFeat.attributes()[inIdx])
-            featList = index.nearestNeighbor(inGeom.asPoint(), nPoints)
             if first:
                 first = False
+                featList = index.nearestNeighbor(inGeom.asPoint(), nPoints)
                 fields = QgsFields()
                 input_id_field = source.fields()[inIdx]
                 input_id_field.setName('ID')
                 fields.append(input_id_field)
-                for i in range(len(featList)):
-                    fields.append(QgsField('DIST_{0}'.format(i + 1), QVariant.Double))
+
+                request = QgsFeatureRequest().setFilterFids(featList).setSubsetOfAttributes([targetIdx])
+                for f in target_source.getFeatures(request):
+                    fields.append(QgsField(f[targetField], QVariant.Double))
+
                 (sink, dest_id) = self.parameterAsSink(parameters, self.OUTPUT, context,
                                                        fields, source.wkbType(), source.sourceCrs())
 
-            data = [inID]
-            for target in target_source.getFeatures(QgsFeatureRequest().setSubsetOfAttributes([]).setFilterFids(featList).setDestinationCrs(source.sourceCrs())):
+            data = [str(inFeat[inField])]
```

Comment:
Is the cast to `str` required?

Reply:
Yes, because unique ID field can be numeric

Reply:
My question was rather, why is it forced to be non-numeric in the output. Sorry if I missed something obvious.

Reply:
Ah sorry, my fault. You are right, this cast is not needed here. Fixed